### PR TITLE
Renderer: Clarified lighting uniforms update.

### PR DESCRIPTION
**Pull request by tschw**
Lighting uniforms are not "refreshed" on the material - they are just wired up by reference.

The GL state is indeed refreshed - and the update is triggered via `.needsUpdate` flags on the uniform (wrapper objects) of the first material interested in lighting.

This implementation is in fact working, but very crude and the code is not sufficiently clear. This PR attempts to shed some light on the subtleties.

Also fixes: 
- Lighting update on camera change.
- Random materials' uniforms getting outdated on current `dev`.

### CODE REVIEW

**Comment by mrdoob** [comment]
Ops! Do you mind doing a new PR using the latest in `dev`?

**Comment by tschw** [comment]
Sure. But `dev` seems momentarily somewhat broken...

**Comment by mrdoob** [comment]
Uhm... I better get on the laptop ☺

**Comment by tschw** [comment]
Interestingly, `dev` also introduces a subtle updating issue because of what this one attempts to tidy up :-).

**Comment by tschw** [comment]
OK, I guess there are just many examples that need updating?


/ping @mrdoob 
Pulled up and added a Bonbon.

**Comment by mrdoob** [comment]
I guess it was the breakage introduced by #7428? Just fixed it.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1570,7 +1585,18 @@ THREE.WebGLRenderer = function ( parameters ) {
 			}
 
 
-			if ( camera !== _currentCamera ) _currentCamera = camera;
+			if ( camera !== _currentCamera ) {
+
+				_currentCamera = camera;
+
+				// lighting uniforms depend on the camera so enforce an update
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I should probably now, but where/how lighting uniforms depend on the camera?

Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Oh, because of the frustum?

**Comment by tschw** [comment]
> I guess it was the breakage introduced by #7428? Just fixed it.

Logarithmic depth buffer and some others are working again.

Still there's flickering. Guess: Something in GLSL does not get initialized. For example:

http://localhost:8000/three.js/examples/#webgl_geometry_colors
http://localhost:8000/three.js/examples/#webgl_geometry_extrude_shapes2
http://localhost:8000/three.js/examples/#webgl_animation_skinning_blending

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (tschw) [inline src/renderers/WebGLRenderer.js]: Because it's done in view space.

**Comment by mrdoob** [comment]
Hmm, I don't see the flickering on my end...?

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Ah, right right.

**Comment by tschw** [comment]
Maybe some drivers zero-initialize and others don't.
window-2015-10-26-16 21 35

**Comment by tschw** [comment]
Forgot to mention: Current `dev`, fresh build, cache disabled. Intel / Mesa Linux driver, Iris GPU.

**Comment by tschw** [comment]
Same with NVIDIA. I see the following warning
```
0(12) : warning C7050: "webgl_3e746cb0e437ba12" might be used before being initialized
```

**Comment by tschw** [comment]
Looks like the varying `vColor` in the vertex shader.

**Comment by tschw** [comment]
/ping @mrdoob 
Problem fixed in this PR. I guess we can merge?

**Comment by mrdoob** [comment]
Thanks!

**Comment by tschw** [comment]
> Interestingly, dev also introduces a subtle updating issue because of what this one attempts to tidy up :-).

Since the lighting uniforms are now set to reference the containing arrays, these must remain the same objects. Otherwise random materials do not receive updates - because the refresh routine is not what it appears to be.

In this PR, I removed that misleading routine and made the wire-up explicit during material initialization. Also, the arrays objects stay the same - and, while at it, their rebuild was optimized to write over the previous content with subsequent truncation instead of emptying with subsequent refill.

**Pull request merged**